sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -42,6 +42,24 @@ abstract class LogicalPlan
    */
   def metadataOutput: Seq[Attribute] = children.flatMap(_.metadataOutput)
 
+  /**
+   * Finds a metadata attribute of this node by its logical name. This search will work even if the
+   * metadata attribute was renamed because of a conflicting name in the data schema.
+   */
+  def getMetadataAttributeByName(name: String): Attribute = {
+    // NOTE: An already-referenced column might appear in `output` instead of `metadataOutput`.
```

Comment:
metadata col may appear in `output`, but always appear in `metadataOutput`. I think `outputMetadataAttributes.resolve(nameParts, resolver)` should do the work.

Reply:
I had an instance in the debugger where this was not the case yesterday

Reply:
Do we have a test to reproduce it? This should not happen after https://github.com/apache/spark/commit/5705436f70e6c6d5a127db7773d3627c8e3d695a

Reply:
`SubqueryAlias` does not always propagate its child's `metadataOutput`, even if the child's own `output` includes metadata columns. That can result in a query plan whose `output` contains attributes which satisfy `MetadataAttribute` but are not in `metadataOutput`. 

I raised https://github.com/apache/spark/pull/40321 as a possible fix for that issue.